### PR TITLE
Compare Models

### PR DESCRIPTION
**Pull request by jshum2479**
This is a draft PR for the model diff - compareModel tool.   It's a work in progress, there are areas that needs more work.  e.g.  do we allow delete certain resource like connection factory, subedeployment ... from JMSSystemResource -> i.e. generate the right notation at the particular level and what's the best way to accomplish that, right now the tool has no concept of aliases and operated solely on dictionary math.

### CODE REVIEW

File: installer/src/main/bin/compareModel.cmd [hunk 1]
```diff
@@ -0,0 +1,73 @@
+@ECHO OFF
+@rem **************************************************************************
+@rem modelHelp.cmd
```
Comment (rakillen) [inline installer/src/main/bin/compareModel.cmd:3]: compareModel.cmd

File: core/src/main/python/model_diff.py [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+        """
+        Update the differences in the final model dictionary with the changes
+        :param ar_changes:   Array of changes in delimited format
+        """
+        # The ar_changes is the keys of changes in the piped format
+        #  'resources|JDBCSystemResource|Generic2|JdbcResource|JDBCConnectionPoolParams|TestConnectionsOnReserve
+        #
+
+        # TODO: may be it can be simplified by walking through levels instead of hard coding?
+
```
Comment (rakillen) [inline core/src/main/python/model_diff.py:211]: I think if you recurse similar to attributes_test.py, you can check for alias_helper.supports_multiple_mbean_instances(), and that will show you which ones have a name level. with apps and libs, I think everything that has a name level can be deleted.

File: installer/src/main/bin/compareModel.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# JAVA_HOME             - The location of the JDK to use.  The caller must set
+#                         this variable to a valid Java 7 (or later) JDK.
+#
+
+
+usage() {
+  echo ""
+  echo "Usage: $1 [-help]"
+  echo "          -oracle_home <oracle_home> "
+  echo "          [-compare_model_output_dir <output_dir> write the outputs to the directory specified]"
```
Comment (rakillen) [inline installer/src/main/bin/compareModel.sh:25]: Maybe shorten to -output_dir ? if we do need to reuse that name in the future, cla_utils can deal with it, similar to checks for -domain_home from different tools.

Opinions?

File: core/src/main/python/model_diff.py [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+        :return:  python dictionary
+        """
+        true = True
+        false = False
+        fh = open(file, 'r')
+        content = fh.read()
+        return eval(content)
+
+
+    def write_dictionary_to_json_file(self, dictionary, writer, indent=''):
```
Comment (rakillen) [inline core/src/main/python/model_diff.py:482]: could PythonToJson in json_translator.py be used for this? might save some duplication.

File: installer/src/main/bin/compareModel.cmd [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+@rem
+@rem     DESCRIPTION
+@rem       This script compares two models. The models compared must be both yaml or both json
+@rem
+@rem This script uses the following variables:
+@rem
+@rem JAVA_HOME             - The location of the JDK to use.  The caller must set
+@rem                         this variable to a valid Java 7 (or later) JDK.
+@rem
+@rem
```
Comment (rakillen) [inline installer/src/main/bin/compareModel.cmd:19]: The variables WLSDEPLOY_HOME and WLSDEPLOY_PROPERTIES are still relevant in jython-only scripts, just copy descriptions from modelHelp.sh and .py

File: core/src/main/python/model_diff.py (hunk 4 above)
Comment (jshum2479) [reply core/src/main/python/model_diff.py]: Yes, fixing it

File: installer/src/main/bin/compareModel.cmd (hunk 5 above)
Comment (jshum2479) [reply installer/src/main/bin/compareModel.cmd:28]: added

File: installer/src/main/bin/compareModel.sh (hunk 3 above)
Comment (jshum2479) [reply installer/src/main/bin/compareModel.sh]: Yes, fixed

File: installer/src/main/bin/compareModel.cmd (hunk 1 above)
Comment (jshum2479) [reply installer/src/main/bin/compareModel.cmd:3]: fixed

File: core/src/main/python/model_diff.py (hunk 2 above)
Comment (jshum2479) [reply core/src/main/python/model_diff.py]: changed to use alias mechanism to walk the tree.

**Comment by jshum2479** [comment]
I added -variable_file option for resolving tokens

File: site/compare.md [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+ 3. Deployed two applications `yourear` and `myear` in the appDeployments section
+ 4. Removed application `myear` in the appDeployments section
+ 4. Removed `MDBConnectionFactory` from the `MyJmsModule` JMS Module in the resource section
+ 5. Removed `WebAppContainer` in the resource section
+ 6. Removed the attribute `ResetDeliveryCountOnForward` from `MyUniformDistributedQueue`
+ 7. Changed deployment targets for application `myear`
+ 8. Removed the attribute 'ModuleType' for application `myear` 
+ 
+ **Note**:  The `!` is a notation for the deletion of a non-attribute key element from the model. Any missing attribute will 
+ be omitted from the resulting model, but shown as messages in the output.
```
Comment (rosemarymarano) [inline site/compare.md:274]: Any missing attribute will be omitted from the resulting model, -> but shown as a message in the output. 
(all singular OR all plural)
Missing attributes will be omitted from the resulting model, but shown as messages in the output.

**Pull request merged**